designs/full-cluster-lifecycle-api.md

```diff
@@ -0,0 +1,224 @@
+
+![Cluster reconcile](images/cluster_reconcile.png)
+
+Note: this diagram is missing things like applying tolerations to a vSphere controller, taints for worker nodes, force updating the `ClusterResourceSet` resources, IAM auth, etc. Those will be added as we start the implementation, when we decide where they should go in the flow. 
+
+
+In order to check if something needs to change, reconcilers are supposed to compare the desired state (in our case, a combination of multiple EKSA objects, like `Cluster`, `<Provider>MachineConfig`, etc.) with the current state (in our case a combination of CAPI objects, CNI and some extra Kubernetes objects, depending on the provider).
+
+Following that, on every reconcile loop we would compare (object by object) if the objects coming from our CAPI "template" are different than the ones already present in the clusters (some of them will be in the management cluster and others in the workload cluster).
+
+This could be an expensive operation, so we could potentially cache it in the status of our cluster object. We would need a combination of all the observed generation's for all the resources that represent our desired state (cluster itself, provider datacenter, machineconfigs...) and the "phase" (for example, "CP reconciled"). A drawback of this approach is that, once the cluster had passed a certain phase, if any of the "result" objects changes (for example by manual user interaction), we would no "reconcile them back" to the desired state. Such reconciliation would only happen upon change in our EKSA objects. We won't implement this right away and will only reconsider if we find performance issues.
+
+Right now object comparison requires a bit of transforming and parsing because our providers' code generates multiobject yaml files. In order to improve this process, we should refactor such code and start using API structs. However, this might or might not be part of the first implementation, depending on the agreed scope.
+
+### `VSphereDatacenterConfig` Validation
+
+In order to maintain some of the validations we run today in the CLI, we will need to port them to the controller. 
+
+Kubernetes offers the validation webhook for CRDs: custom code to validate objects before been accepted by the kube-api server. However, these validation are supposed to be light and fast. Most our validations require calling the vSphere API, which makes them too heavy for the validation webhook.
+
+We will implement those validations in a separate reconciler. This means that they will run after the object has already been accepted by the api server. We will store if the object has been validated or not as well as the validation error (if any) in the V`SphereDatacenterConfig` status.
```

Comment:
```suggestion
We will implement those validations in a separate reconciler. This means that they will run after the object has already been accepted by the api server. We will store if the object has been validated or not as well as the validation error (if any) in the `VSphereDatacenterConfig` status.
```

Reply:
So the status will have a "Validated" state?

Reply:
@vivek-koppuru yeah!